### PR TITLE
Fix warnings in release mode

### PR DESCRIPTION
**Pull request by Rombur**
I get warnings when compiling ArborX in release mode with g++ 11.2. I don't get any warnings in debug mode. The first commit is pretty uncontroversial, I am not sure if we want the second one. It looks like a false positive.

### CODE REVIEW

**Comment by dalg24** [comment]
Where did you see the `-Warray-bounds` warning?  Was it when building ArborX tests/examples?
I am tempted to not do anything yet about that one

**Comment by Rombur** [comment]
The warning appears when we use `ordered_intersect` with rays.

**Comment by masterleinad** [comment]
Why are we getting the `array-bounds` warning? What does the compiler print?

**Comment by Rombur** [comment]
I don't know why we get the warning but it's from https://github.com/arborx/ArborX/blob/master/src/details/ArborX_DetailsTreeTraversal.hpp#L505-L515

```
In file included from /home/ArborX/src/details/ArborX_DetailsPriorityQueue.hpp:15,                                                                                                                                 
                 from /home/ArborX/src/details/ArborX_DetailsDistributedTreeImpl.hpp:21,                                                                                                                           
                 from /home/ArborX/src/ArborX_DistributedTree.hpp:16,                                                                                                                                              
                 from /home/ArborX/src/ArborX.hpp:19,                                                                                                                                                              
                 from /home/ArborX/examples/raytracing/example_raytracing.cpp:25:                                                                                                                                  
/home/ArborX/src/details/ArborX_DetailsHeap.hpp: In member function 'void ArborX::Details::TreeTraversal<BVH, Predicates, Callback, ArborX::Experimental::OrderedSpatialPredicateTag>::operator()(int) const [with 
BVH = ArborX::BasicBoundingVolumeHierarchy<Kokkos::HostSpace, ArborX::Box, void>; Predicates = ArborX::Details::PermutedData<OrderedIntersectsBased::Rays<Kokkos::HostSpace>, Kokkos::View<unsigned int*, Kokkos::D
evice<Kokkos::OpenMP, Kokkos::HostSpace> >, false>; Callback = OrderedIntersectsBased::DepositEnergy<Kokkos::HostSpace>]':                                                                                         
/home/ArborX/src/details/ArborX_DetailsHeap.hpp:105:5: warning: array subscript -1 is outside array bounds of 'PairIndexDistance [64]' {aka 'Kokkos::pair<int, float> [64]'} [-Warray-bounds]                      
  105 |     *(last - 1) = std::move(value);                                                                                                                                                                        
      |     ^                                                                                                                                                                                                      
In file included from /home/ArborX/src/ArborX_LinearBVH.hpp:26,                                                                                                                                                    
                 from /home/ArborX/src/details/ArborX_DetailsDistributedTreeImpl.hpp:23,                                                                                                                           
                 from /home/ArborX/src/ArborX_DistributedTree.hpp:16,                                                                                                                                              
                 from /home/ArborX/src/ArborX.hpp:19,                                                                                                                                                              
                 from /home/ArborX/examples/raytracing/example_raytracing.cpp:25:                                                                                                                                  
/home/ArborX/src/details/ArborX_DetailsTreeTraversal.hpp:460:23: note: while referencing 'buffer'                                                                                                                  
  460 |     PairIndexDistance buffer[buffer_size];                                                                                                                                                                 
      |                       ^~~~~~               
```

**Comment by aprokop** [comment]
Yeah, the warning seems off.

File: src/details/ArborX_DetailsHeap.hpp [hunk 1]
```diff
@@ -102,7 +105,10 @@ KOKKOS_INLINE_FUNCTION void popHeap(RandomIterator first, RandomIterator last,
     ValueType value = std::move(*first);
     bubbleDown(first, DistanceType(0), DistanceType((last - first) - 1),
                std::move(*(last - 1)), comp);
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Warray-bounds"
     *(last - 1) = std::move(value);
+#pragma GCC diagnostic pop
```
Comment (aprokop) [inline src/details/ArborX_DetailsHeap.hpp:111]: This is peculiar. It does not complain on line 106, but is complaining on line 110? And this is despite using the same `*(last - 1)`.

**Comment by aprokop** [comment]
@Rombur Can you please try this patch:
```diff
--- a/src/details/ArborX_DetailsHeap.hpp
+++ b/src/details/ArborX_DetailsHeap.hpp
@@ -60,11 +60,12 @@ KOKKOS_INLINE_FUNCTION void pushHeap(RandomIterator first, RandomIterator last,
   using DistanceType =
       typename std::iterator_traits<RandomIterator>::difference_type;
   using ValueType = typename std::iterator_traits<RandomIterator>::value_type;
-  if (last - first > 1)
+  auto const n = last - first;
+  if (n > 1)
   {
-    ValueType value = std::move(*(last - 1));
-    bubbleUp(first, DistanceType((last - first) - 1), DistanceType(0),
-             std::move(value), comp);
+    ValueType value = std::move(*(first + n - 1));
+    bubbleUp(first, DistanceType(n - 1), DistanceType(0), std::move(value),
+             comp);
   }
 }

@@ -97,12 +98,13 @@ KOKKOS_INLINE_FUNCTION void popHeap(RandomIterator first, RandomIterator last,
   using DistanceType =
       typename std::iterator_traits<RandomIterator>::difference_type;
   using ValueType = typename std::iterator_traits<RandomIterator>::value_type;
-  if (last - first > 1)
+  auto const n = last - first;
+  if (n > 1)
   {
     ValueType value = std::move(*first);
-    bubbleDown(first, DistanceType(0), DistanceType((last - first) - 1),
-               std::move(*(last - 1)), comp);
-    *(last - 1) = std::move(value);
+    bubbleDown(first, DistanceType(0), DistanceType(n - 1),
+               std::move(*(first + n - 1)), comp);
+    *(first + n - 1) = std::move(value);
   }
 }
```

**Comment by Rombur** [comment]
@aprokop your fix works

**Review by aprokop — APPROVED** [review]

**Review by masterleinad — APPROVED** [review]
Looks OK to me.

**Pull request merged**